[ENH] Increase configurability of API/graph server parameters

alyssadai opened the pull request:
<!--- Until this PR is ready for review, you can include the [WIP] tag in its title, or create a draft PR. -->


<!---
Below is a suggested pull request template.
It's designed to capture info we've found to be useful in reviewing pull requests, but feel free to add more details you feel are relevant/necessary.
-->

<!-- 
Please indicate after the # which issue you're closing with this PR, if applicable.
If the PR closes multiple issues, include "closes" before each one is listed.
You can also link to other issues if necessary, e.g. "See also #1234".

https://help.github.com/articles/closing-issues-using-keywords
-->
Closes #118
Closes #122
Closes #140 
Closes #142
Closes https://github.com/neurobagel/project/issues/162

<!-- 
Please give a brief overview of what has changed or been added in the PR.
This can include anything specific the maintainers should be looking for when they review the PR.
-->
Changes proposed in this pull request:

- Make the following configurable as environment variables:
  - host + container ports
  - graph server Docker image
  - graph root paths on host + container
  - full graph database query endpoint
- Update README with new env vars + simplified instructions for using .env file when using building or using API Docker image alone
- Increase timeout for query to minimize risk of occasional timeouts
- Add an annotated template .env file

<!-- To be checked off by reviewers -->
## Checklist

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[FIX]`, `[REF]`, `[TST]`, `[CI]`, `[MNT]`, `[INF]`, `[MODEL]`, `[DOC]`) _(see https://neurobagel.org/contributing/pull_requests for more info)_
- [x] PR links to GitHub issue with mention `Closes #XXXX`
- [x] Tests pass
- [x] Checks pass

For new features:
- [ ] Tests have been added

For bug fixes:
- [ ] There is at least one test that would fail under the original bug conditions.

surchs commented on Dockerfile:
```diff
@@ -8,6 +8,4 @@ RUN pip install --no-cache-dir --upgrade -r /usr/src/app/requirements.txt
-
-CMD ["--port", "8000"]
\ No newline at end of file
+ENTRYPOINT uvicorn app.main:app --proxy-headers --host 0.0.0.0 --port ${NB_API_PORT:-8000}
```
1. I feel like the removal of `CMD` keeps showing up in some PRs but I don't think we want that, no? [`CMD`](https://docs.docker.com/engine/reference/builder/#cmd) gives me a default I can easily change. [`ENTRYPOINT`](https://docs.docker.com/engine/reference/builder/#entrypoint) gives me a default I shouldn't have to change. Port is probably something we want to have as a CMD default then
2. `${NB_API_PORT:-8000}` this looks cool - but how does it get in here? I would have expected some `ARG` or `ENV` command to set it first? More importantly: what is it supposed to do? Let you change the default port at build time?
3. I'm assuming there is a reason you changed this line from the `exec` form to the `shell` form - is this linked to the appearance of the shell variable and the disappearance of the CMD?

surchs commented on README.md:
```diff
@@ -56,60 +56,71 @@ NB_GRAPH_USERNAME=someuser
+| `NB_RETURN_AGG`      | No                | Whether to return only dataset-level query results (including data locations) and exclude subject-level attributes. One of [true, false] | `true`                                 | Docker, Python                |
+| `NB_API_TAG`         | No                | Tag for API Docker image                                                                                                                 | `latest`                               | Docker                        |
+| `NB_API_PORT_HOST`   | No                | Port number on the _host machine_ to map the API container port to                                                                       | `8000`                                 | Docker, Python                |
+| `NB_API_PORT`        | No                | Port number to use for the _API container_                                                                                               | `8000`                                 | Docker, Python                |
```
```suggestion
| `NB_API_PORT`        | No                | Port number to expose on the _API container_                                                                                               | `8000`                                 | Docker, Python                |
```
does that fit better?

surchs commented on README.md:
```diff
@@ -56,60 +56,71 @@ NB_GRAPH_USERNAME=someuser
+| `NB_GRAPH_PORT_HOST` | No                | Port number on the _host machine_ to map the graph server container port to                                                              | `5820`                                 | Docker                        |
+| `NB_GRAPH_PORT`      | No                | Port number used by the _graph server container_                                                                                         | `5820` *                               | Docker                        |
+
+_* These defaults should not be changed if using a Stardog backend._
```
```suggestion
_* These defaults are configured for a Stardog backend - you should not have to change them if you are running a Stardog backend._
```

surchs commented on README.md:
```diff
@@ -56,60 +56,71 @@ NB_GRAPH_USERNAME=someuser
 
 If needed, update your `.env` file with optional environment variables for the docker compose configuration.
 
+**TIP:** Double check that the environment variables are resolved with your expected values using the command `docker compose config`.
```
nice!

surchs commented on README.md:
```diff
@@ -56,60 +56,71 @@ NB_GRAPH_USERNAME=someuser
+export $(cat .env | xargs)  # OPTIONAL: export your environment variables
 docker pull neurobagel/api
-docker run -d --name api -p 8000:8000 --env NB_GRAPH_USERNAME --env NB_GRAPH_PASSWORD neurobagel/api
+docker run -d --name=api -p ${NB_API_PORT_HOST:-8000}:${NB_API_PORT:-8000} --env-file=.env neurobagel/api
```
I get what you have in mind here - but I don't think the README section for "minimal command to run this" is the right place for this relatively complex syntax. As far as I understand, the motivation for `NB_API_PORT_HOST` and `NB_API_PORT` is when I use `docker compose`, yes? in that case I'd just stick with a literal port value - we can still mention that this can be changed.

surchs commented on app/api/crud.py:
```diff
@@ -85,6 +85,7 @@ async def get(
                 os.environ.get(util.GRAPH_USERNAME.name),
                 os.environ.get(util.GRAPH_PASSWORD.name),
             ),
+            timeout=10.0,
```
1. I think this means you can also close https://github.com/neurobagel/api/issues/140 with this
2. I also think that 10 seconds is not enough given the query times we have already seen - until we fix the underlying issue of unoptimized and slow queries, let's not set this value too low. How about a) a TODO comment to revisit the value when performance improves, b) 30 seconds?

surchs reviewed (changes requested):
Thanks for addressing this @alyssadai. I think this is good to go, but I have not yet approved it because I have some questions about the Dockerfile and the API port setting changes you made. It's possible (likely?) that I just don't understand all you are doing here but I think it'd be good to make sure first. 

Take a look

alyssadai commented:
Hey @surchs, thanks again for your detailed review of this.

In the README, I updated the `docker run` commands/corresponding instructions in the option 2 and option 3 subsections of the Docker installation section based on your comments. I also added a note to the Dockerfile about how the environment variable for the API port behaves.

Finally, I've added a commented template .env file and instructions to use it in the README.
